Add bare metal support to NXP targets

harmut01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes https://github.com/ARMmbed/mbed-os-example-blinky-baremetal/issues/55

Modifies`targets.json` to provide bare metal support to the following targets: LPC1768, LPC54114, and LPC546XX. The corresponding scatter files have also been changed to ensure that a heap load region exists. This pull request also removes LPC1114 µARM dependencies as this tool-chain has been deprecated.   

Issue raised in #13099.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
None
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
None
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
None
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

harrisonmutai-arm committed:
Remove µARM toolchain dependencies from NXP targets

harrisonmutai-arm committed:
Add bare metal support to NXP targets

Modify linker scripts to add heap to memory map, add list of supported c libs and "bare-metal" to supported_application_profiles.

hugueskamba commented on targets/TARGET_NXP/TARGET_MCUXpresso_MCUS/TARGET_LPC54114/device/TARGET_LPC54114_M4/TOOLCHAIN_ARM_STD/LPC54114J256_cm4.sct:
```diff
@@ -121,6 +121,8 @@ LR_IROM1 m_interrupts_start m_text_start+m_text_size-m_interrupts_start { ; load
   }
   RW_IRAM1 ((ImageLimit(RW_m_data) == m_data_start) ? ImageLimit(RW_m_data) : +0) EMPTY Heap_Size { ; Heap region growing up
   }
+  ARM_LIB_HEAP AlignExpr(+0, 16) EMPTY Heap_Size{ ; Heap region growing up
```
`Heap_Size` should probably be calculated as done for other platforms (`MBED_RAM_SIZE - RAM_FIXED_SIZE - (AlignExpr(ImageLimit(RW_IRAM1), 16) - MBED_RAM1_START)`) instead of setting it to `0x0400`.
@ARMmbed/team-nxp can you confirm?

harmut01 replied:
@hugueskamba I've had to remove the dynamic calculation of the heap size since it was causing CI to fail. The change I made resulted in a circular definition between Heap_Size and RW_IRAM1.

hugueskamba replied:
Can you show me what you had that caused the circular definition?

harmut01 replied:
This was the original change:
```diff
+ #define Heap_Size                    (m_data_size - Stack_Size - (AlignExpr(ImageLimit(RW_IRAM1), 16) - m_data_start))
- #define Heap_Size                    0x0400
```

hugueskamba replied:
@ARMmbed/team-nxp any thoughts?

hugueskamba commented on targets/targets.json:
```diff
@@ -345,7 +343,19 @@
+                "std", "small"
+            ]
+        },
+        "c_lib": "small",
```
This should only be here if it is a former Mbed 2 target or a target that is ultra-constrained (which I do not think it is). @rajkan01 Can you confirm?

rajkan01 replied:
Yes target config "c_lib": _"small"_ should be there for the Mbed 2 target

hugueskamba commented on targets/targets.json:
```diff
@@ -4293,7 +4303,19 @@
+                "std", "small"
+            ]
+        },
+        "c_lib": "small",
```
Same comment as above.

rajkan01 replied:
The target config "c_lib": "small" should be there for the Mbed 2 target

hugueskamba replied:
@harmut01 Please remove `"c_lib": "small",`.

hugueskamba commented on targets/targets.json:
```diff
@@ -4347,7 +4369,19 @@
+                "std", "small"
+            ]
+        },
+        "c_lib": "small",
```
Same comment as above.

rajkan01 replied:
The target config "c_lib": "small" should be there for the Mbed 2 target

hugueskamba replied:
@harmut01 Please remove `"c_lib": "small",`.

harrisonmutai-arm committed:
Remove 'c_lib' configuration from NXP targets

hugueskamba commented on targets/targets.json:
```diff
@@ -295,9 +295,7 @@
-            "uARM",
-            "GCC_ARM",
-            "IAR"
+            "GCC_ARM"
```
Please do the same for all NXP targets (perhaps as a separate commit). There are inconsistencies where IAR is listed as a supported toolchain but no supported c libs are listed for IAR. See LPC541114 for example.

harrisonmutai-arm committed:
Remove IAR from NXP targets without small c lib support

Toolchains should provide support for small c lib, remove IAR from the
NXP targets that don't currently do so.

LDong-Arm commented on targets/targets.json:
```diff
@@ -345,7 +343,18 @@
+            "full", "bare-metal"
+        ]
     },
     "ARCH_PRO": {
```
Please add
```
        "supported_application_profiles": [
            "full", "bare-metal"
        ]
```
to this target. It's required in order to fix https://github.com/ARMmbed/mbed-os-example-blinky-baremetal/issues/55.

LDong-Arm reviewed (changes requested):
We should aim to get this PR tested and merged after the review comment is addressed.

harrisonmutai-arm committed:
Add bare metal support to ARCH_PRO

LDong-Arm reviewed (approved):
LGTM
